crypto/decrypting_reader.go

```diff
@@ -0,0 +1,134 @@
+package crypto
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/cipher"
+	"errors"
+	"io"
+)
+
+func newBlockModeDecryptingReader(r io.Reader, mode cipher.BlockMode) io.Reader {
+	return &blockModeDecryptingReader{
+		r:         bufio.NewReader(r),
+		blockMode: mode,
+		buffer:    bytes.NewBuffer(nil),
+		err:       nil,
+	}
+}
+
+type blockModeDecryptingReader struct {
+	r         *bufio.Reader
+	blockMode cipher.BlockMode
+	buffer    *bytes.Buffer
+	err       error
+}
+
+func (decryptingReader *blockModeDecryptingReader) readNextBlock() ([]byte, error) {
+	reader := decryptingReader.r
+
+	output := make([]byte, decryptingReader.blockMode.BlockSize())
+	sizeOfCurrentlyRead, readErr := io.ReadFull(reader, output)
+	if readErr != nil && readErr != io.EOF {
+		return output[:sizeOfCurrentlyRead], readErr
+	}
+
+	if sizeOfCurrentlyRead == 0 && readErr == io.EOF {
+		return nil, io.EOF
+	}
+
+	if _, e := reader.Peek(1); e == io.EOF {
+		return output[:sizeOfCurrentlyRead], io.EOF
+	}
+
+	return output, nil
+}
+
+func (decryptingReader *blockModeDecryptingReader) decryptUntilPFull(p []byte) (n int, err error) {
```

Comment:
as above - should we return anything in case of error?
Is having any return (even info about how much data have been read) viable for users?

Reply:
As above. I'll check it and let you know if I have problems